File: Fortran/02_Types/02_Indexed/indexed.f90
PROGRAM type

  USE mpi
  IMPLICIT NONE

  INTEGER, PARAMETER :: tag = 100
  INTEGER, PARAMETER :: nitems = 8

  INTEGER :: rank, recv_rank
  INTEGER :: nproc
  INTEGER :: left, right
  INTEGER :: ierr
  INTEGER :: index_type
  INTEGER, DIMENSION(nitems) :: values, values_recv
  INTEGER, DIMENSION(2) :: lengths, displacements

  CALL MPI_Init(ierr)

  CALL MPI_Comm_size(MPI_COMM_WORLD, nproc, ierr)
  CALL MPI_Comm_rank(MPI_COMM_WORLD, rank, ierr)

  !Set up periodic domain
  left = rank - 1
  IF (left < 0) left = nproc - 1
  right = rank + 1
  IF (right > nproc - 1) right = 0

  values = rank
  values_recv = 0

  lengths = (/, /)
  displacements = (/, /)
  !Create the type
  CALL MPI_Type_indexed(2, lengths, displacements, MPI_INTEGER, index_type, &
      ierr)

  !Register the type
  CALL MPI_Type_commit(index_type, ierr)

  IF (rank == 0) THEN
    PRINT *,'MPI_Type_indexed used as send and recieve types'
  END IF
  !Sendrecv using the type. Note that sendcount and recvcount are now both 1
  CALL MPI_Sendrecv(values, 1, index_type, right, tag, values_recv, &
      1, index_type, left, tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE, ierr)

  IF (rank == 0) THEN
    PRINT *, 'Detailed result on rank 0 is ', values_recv
  END IF

  CALL MPI_Finalize(ierr)

END PROGRAM type

File: C/02_Types/02_Indexed/indexed.c
#include <stdio.h>
#include <mpi.h>

#define TAG 100
#define NITEMS 8

int main(int argc, char** argv)
{

  int rank, recv_rank, nproc, left, right, iitems, sum;
  int values[NITEMS], values_recv[NITEMS];
  int lengths[2], displacements[2];
  MPI_Datatype index_type;

  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);

  //Set up periodic domain
  left = rank - 1;
  if (left < 0) left = nproc - 1;
  right = rank + 1;
  if (right > nproc - 1) right = 0;

  for (iitems = 0; iitems < NITEMS; ++iitems){
    values[iitems] = rank;
    values_recv[iitems] = 0;
  }

  lengths[0] = ; lengths[1] = ;
  displacements[0] = ; displacements[1] = ;

  MPI_Type_indexed(2, lengths, displacements, MPI_INT, &index_type);

  MPI_Type_commit(&index_type);

  if (rank == 0) printf("MPI_Type_indexed used as send and recieve types\n");

  MPI_Sendrecv(values, 1, index_type, right, TAG, values_recv, 1,
      index_type, left, TAG, MPI_COMM_WORLD, MPI_STATUS_IGNORE);

  MPI_Type_free(&index_type);

  if (rank == 0) {
    printf("Detailed result on rank 0 is ");
    for (iitems = 0; iitems < NITEMS; ++iitems){
      printf("%3d ", values_recv[iitems]);
    }
    printf("\n");
  }

  MPI_Finalize();

}
